### PR TITLE
Add support for Python 3.8

### LINKED ISSUE #4311: Add support for Python 3.8
This will be more effort than for Python 3.7 since the Python initialization sequence has been changes and the bootloader needs quite some changes.

Anybody going to [fund this development](https://www.pyinstaller.org/funding.html)? Otherwise it might be a long time until this will happen.

### LINKED ISSUE #4865: Reoccurrence of #2322? Multiprocessing spawn in python 3 does not work on macOS
## Description of the issue
There looks to be a reoccurrence of #2322 on macOS with spawn + multiprocessing. I came across this issue with a package that I develop (that uses PyQt5). When I try and use multiprocessing within my program I get multiple GUI windows displaying. This has only happened since building the executable with Py3.8 (spawn is default), it doesn't happen with Py3.7 (fork is default). I tried using the MWE from #2322 and it is failing on my computer.

## Context information

macOS 10.13.6
Python 3.8.2 (from conda)
PyInstaller: tip of the development branch (3.6)

  * [Y] start with clean installation
  * [Y] use the latest development version
  * [Y] Run your frozen program **from a command window (shell)** — instead of double-clicking on it
  * [Y] Package your program in **--onedir mode**
  * [Y] Package **without UPX**, say: use the option `--noupx` or set `upx=False` in your .spec-file
  * [Y] Repackage you application in **verbose/debug mode**. For this, pass the option `--debug` to `pyi-makespec` or `pyinstaller` or use `EXE(..., debug=1, ...)` in your .spec file.

### A minimal example program which shows the error
Running `test.py` from bash is fine, and gives the expected result. However, running the packaged executable results in processes spawning continuously.

**test.py**
```
import sys
from multiprocessing import freeze_support, set_start_method, Process

def test():
    print('In subprocess')

if __name__ == '__main__':
    print(sys.argv,)
    freeze_support()
    set_start_method('spawn')

    print('In main')
    proc = Process(target=test)
    proc.start()
    proc.join()
```
**test.spec**
```
# -*- mode: python -*-

block_cipher = None


a = Analysis(['test.py'],
             binaries=[],
             datas=[],
             hiddenimports=[],
             hookspath=[],
             runtime_hooks=[],
             excludes=[],
             win_no_prefer_redirects=False,
             win_private_assemblies=False,
             cipher=block_cipher)
pyz = PYZ(a.pure, a.zipped_data,
             cipher=block_cipher)
exe = EXE(pyz,
          a.scripts,
          exclude_binaries=True,
          name='test',
          debug=True,
          strip=False,
          upx=True,
          console=True )
coll = COLLECT(exe,
               a.binaries,
               a.zipfiles,
               a.datas,
               strip=False,
               upx=True,
               name='test')
```
### Stacktrace / full error message
Log at https://gist.github.com/andyfaff/c8402452e7bf4d27a1c86c1c486a736e

### PR DESCRIPTION
**Pull request by alexchandel**
This attempt adds initial support for Python 3.8.  Closes #4311.  Changes include:

* add `compat.is_py38` flag
* add `code.co_posonlyargcount` to `modulegraph`
* fix multiprocessing spawn in 3.8 (fixes #4865)
* fix PyiModuleGraph test, due to import changes in builtin modules
* CI and small documentation updates

While Python 3.8 introduced a brand-new configuration API for the bootloader, [investigation indicates](https://github.com/pyinstaller/pyinstaller/issues/4311#issuecomment-652633894) that the new API remains optional for 3.8.

### CODE REVIEW

**Comment by alexchandel** [comment]
Test dependency `matplotlib==3.0.3` causes the Windows Py3.8 tests to fail, as no wheel exists, and building from source fails due to missing `freetype` & `png` libraries.

**Comment by alexchandel** [comment]
Anyone have any idea on [this Windows error](https://ci.appveyor.com/project/PyInstaller/pyinstaller/builds/34017359/job/o896virfwxv7hm5o)?  It's `test_pywin32_win32ui` / `test_option_runtime_tmpdir` / `test_spec_with_multipackage`.  Latter 2 tests pass on other platforms.  cc @Legorooj @htgoebel

File: .github/ISSUE_TEMPLATE/bug_report.md [hunk 1]
```diff
@@ -25,7 +25,7 @@ Welcome to the PyInstaller issue tracker! Before creating an issue, please heed
 ## Context information (for bug reports)
 
 * Output of `pyinstaller --version`: ```(paste here)```
-* Version of Python: e.g. 3.7
+* Version of Python: e.g. 3.8
```
Comment (Legorooj) [inline .github/ISSUE_TEMPLATE/bug_report.md:28]: There is no point changing this; 3.7 is still a valid version. Please remove this change.

File: .travis.yml [hunk 2]
```diff
@@ -153,7 +160,7 @@ addons:
   homebrew:
     packages:
     - python
-    # Update to Python 3.7 via homebrew.
+    # Update to Python 3.8 via homebrew.
```
Comment (Legorooj) [inline .travis.yml:163]: Please also remove this change.

File: PyInstaller/hooks/rthooks/pyi_rth_multiprocessing.py [hunk 3]
```diff
@@ -86,3 +92,27 @@ def __init__(self, *args, **kw):
...
+
+# fix spawn as well
+# Patch Popen to re-set _MEIPASS2 from sys._MEIPASS.
+if not sys.platform.startswith('win'):
+    class _Spawning_Popen(spawning.Popen):
+        def __init__(self, *args, **kw):
+            if hasattr(sys, 'frozen'):
+                # We have to set original _MEIPASS2 value from sys._MEIPASS
+                # to get --onefile mode working.
+                os.putenv('_MEIPASS2', sys._MEIPASS)  # @UndefinedVariable
```
Comment (Legorooj) [inline PyInstaller/hooks/rthooks/pyi_rth_multiprocessing.py:104]: Please remove the linter comment.
```suggestion
                os.putenv('_MEIPASS2', sys._MEIPASS) 
```

File: PyInstaller/lib/modulegraph/modulegraph.py [hunk 4]
```diff
@@ -3394,7 +3394,15 @@ def _replace_paths_in_code(self, co):
...
-        if hasattr(co, 'co_kwonlyargcount'):
+        if hasattr(co, 'co_posonlyargcount'):  # >= 3.8
+            return code_func(
+                co.co_argcount, co.co_posonlyargcount,
+                co.co_kwonlyargcount, co.co_nlocals,
+                co.co_stacksize, co.co_flags, co.co_code,
+                tuple(consts), co.co_names, co.co_varnames,
+                new_filename, co.co_name, co.co_firstlineno,
+                co.co_lnotab, co.co_freevars, co.co_cellvars)
+        elif hasattr(co, 'co_kwonlyargcount'):
```
Comment (Legorooj) [inline PyInstaller/lib/modulegraph/modulegraph.py:3405]: Thanks for these modulegraph changes - leave them for now, though you'll have to remove them later once I port them onto my provate modulegraph move branch. The new modulegraph is radically different and I'm keeping the work local for now.

File: doc/pyinstaller.1 [hunk 5]
```diff
@@ -43,7 +43,7 @@ PyInstaller is a program that freezes (packages) Python programs into
 stand\-alone executables, under Windows, GNU/Linux, Mac OS X,
 FreeBSD, OpenBSD, Solaris and AIX.
 Its main advantages over similar tools are that PyInstaller works with
-Python 2.7 and 3.5—3.7, it builds smaller executables thanks to transparent
+Python 2.7 and 3.5—3.8, it builds smaller executables thanks to transparent
```
Comment (Legorooj) [inline doc/pyinstaller.1:46]: Please remove this change for now - the man pages need syncing anyway, and we can add this in when that get's done.

File: tests/requirements-libraries.txt [hunk 6]
```diff
@@ -18,7 +18,8 @@
 # - v. 2.2 and above fails.
 Django==2.1.8  # pyup: ignore
 # - v 3.1.0 fails.
-matplotlib==3.0.3  # pyup: ignore
+matplotlib==3.0.3; python_version < '3.8'  # pyup: ignore
+matplotlib==3.2.2; python_version >= '3.8'  # pyup: ignore
```
Comment (Legorooj) [inline tests/requirements-libraries.txt:22]: The matplotlib hooks need updating so you'll have to change this soon. I hope I can get this done today.

File: tests/scripts/Dockerfile [hunk 7]
```diff
@@ -8,7 +8,7 @@
 # by the PyInstaller maintainers. If you find it useful,
 # please consider contributing to its maintenance and improvement.
 
-FROM python:3.7
+FROM python:3.8
```
Comment (Legorooj) [inline tests/scripts/Dockerfile:11]: Please remove this change.

File: .github/ISSUE_TEMPLATE/bug_report.md (hunk 1 above)
Comment (alexchandel) [reply .github/ISSUE_TEMPLATE/bug_report.md]: The point was just to have the most recent version defaulted into the template, as that's where most bugs would surface.

File: tests/requirements-libraries.txt (hunk 6 above)
Comment (Legorooj) [reply tests/requirements-libraries.txt:22]: Ok after an unexpected work item came up I'm not going to get this done today. Hopefully in the next few days though!

**Comment by alexchandel** [comment]
Here are logs for two failing tests (the other 3 are the same): [test_basic.log](https://github.com/pyinstaller/pyinstaller/files/4915557/test_basic.log)

The failure is a DLL error importing any `pywin32` extension module.  (Note other extension modules like `_tkinter` freeze fine.)  The unimportable DLL modules all depend on `pywintypes38.dll`.  And the imports fail because `pywintypes38.dll` is never frozen into the archive.

But the one-line programs `import pywintypes` (loads `pywintypes38.dll`), `import pythoncom` (imports `pywintypes` and its DLL), `import win32com` (imports `win32api` and `pythoncom`), and `import win32serviceutil` (imports `win32api` and `pywintypes`) freeze successfully.

So the issue is, pyinstaller fails to find `pywintypes38.dll` (or `pythoncom38.dll`) unless it's explicitly imported; DLL dependencies aren't enough (yielding `WARNING  PyInstaller.depend.bindepend:bindepend.py:543 lib not found: pywintypes38.dll dependency of c:\users\user\desktop\pyinstaller\venv\lib\site-packages\win32\win32api.pyd`).

Why is this?  An identical 3.7 venv succeeds, with DLLs in `site-packages\pywin32_system32` just like 3.8.  How does 3.7 find and add the DLL, but 3.8 doesn't?

File: tests/requirements-libraries.txt (hunk 6 above)
Comment (alexchandel) [reply tests/requirements-libraries.txt:22]: The changes address the `3.1.0` incompatibility, stemming from `rcParams`.

File: PyInstaller/lib/modulegraph/modulegraph.py (hunk 4 above)
Comment (alexchandel) [reply PyInstaller/lib/modulegraph/modulegraph.py:3405]: Then we can just merge these changes in, and leave them until modulegraph is replaced with moduelgraph2, right?

**Comment by Legorooj** [comment]
@alexchandel what is the difference between the installation. Is there a difference? (The folders in site-packages I mean)

**Comment by alexchandel** [comment]
@Legorooj Completely identical.  Both created like `pip install -r requirements; pip install pywin32`.  Running the following frozen code fails for 3.8, works for 3.7:

```py
import win32api; print(win32api)
```

**Comment by alexchandel** [comment]
@Legorooj It's because `PyInstaller.depend.bindepend.getfullnameof` can't resolve `pywintypes38.dll`.  I've narrowed this down to the fact that Python 3.7 adds `pywin32_system32` to its PATH, but `3.8` doesn't:

```bat
(v37) C:\Users\User\Desktop\pyinstaller>path
PATH=C:\Users\User\Desktop\pyinstaller\venvs\v37\Scripts;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Program Files\dotnet\;C:\Program Files\Microsoft SQL Server\130\Tools\Binn\;C:\Program Files\Microsoft SQL Server\Client SDK\ODBC\170\Tools\Binn\;C:\Users\User\AppData\Local\Microsoft\WindowsApps;
(v37) C:\Users\User\Desktop\pyinstaller>python -c "import os; print(os.environ['PATH'])"
C:\Users\User\Desktop\pyinstaller\venvs\v37\lib\site-packages\pywin32_system32;C:\Users\User\Desktop\pyinstaller\venvs\v37\Scripts;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Program Files\dotnet\;C:\Program Files\Microsoft SQL Server\130\Tools\Binn\;C:\Program Files\Microsoft SQL Server\Client SDK\ODBC\170\Tools\Binn\;C:\Users\User\AppData\Local\Microsoft\WindowsApps;
```

```bat
(v38) C:\Users\User\Desktop\pyinstaller>path
PATH=C:\Users\User\Desktop\pyinstaller\venvs\v38\Scripts;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Program Files\dotnet\;C:\Program Files\Microsoft SQL Server\130\Tools\Binn\;C:\Program Files\Microsoft SQL Server\Client SDK\ODBC\170\Tools\Binn\;C:\Users\User\AppData\Local\Microsoft\WindowsApps;
(v38) C:\Users\User\Desktop\pyinstaller>python -c "import os; print(os.environ['PATH'])"
C:\Users\User\Desktop\pyinstaller\venvs\v38\Scripts;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Program Files\dotnet\;C:\Program Files\Microsoft SQL Server\130\Tools\Binn\;C:\Program Files\Microsoft SQL Server\Client SDK\ODBC\170\Tools\Binn\;C:\Users\User\AppData\Local\Microsoft\WindowsApps;
```

**Comment by alexchandel** [comment]
@Legorooj This is because `pywin32_bootstrap` (called by `pywin32.pth`) calls now calls [os.add_dll_directory()](https://docs.python.org/3.8/library/os.html#os.add_dll_directory) (which calls `Kernel32.dll!AddDllDirectory`) rather than prepending `os.environ["PATH"]` in Python 3.8.  There is no mechanism in Windows to list directories added in this way, and [os.add_dll_directory() doesn't record it](https://github.com/python/cpython/blob/3.8/Lib/os.py#L1098).

**Comment by alexchandel** [comment]
I've duplicated the `numpy/core` hack, which fixes the build.

**Comment by alexchandel** [comment]
@Legorooj does your linter ban [W503](https://travis-ci.org/github/pyinstaller/pyinstaller/jobs/708141780#L1041) *and* W504?  What is one supposed to do?

**Comment by alexchandel** [comment]
Log for `test_spec_with_multipackage[onedir_depends_on_onedir]`: 
[test_multipackage.onedir.log](https://github.com/pyinstaller/pyinstaller/files/4927631/test_multipackage.onedir.log) Any ideas?  I checked, the `multipackage4_B` folder exists and has everything.

**Comment by LeighKorbel** [comment]
@Legorooj suggested I submit a pull request for Python 3.8 support for issue https://github.com/pyinstaller/pyinstaller/issues/4996, but this is unfortunately something outside of the scope of my abilities so I will simply comment about it here since it looks like additional Python 3.8 support is already being done here by @alexchandel . All of my attempts to install PyInstaller in MSYS2 have failed with:

`Your platform is not yet supported. Please define constant PYDYLIB_NAMES for your platform.`

@Legorooj stated:

> We already semi-support MSYS2, except that we aren't testing for whether we're on MSYS2 properly it appears. Please submit a pull request that modifies compat.py so it uses the Windows PYDYLIB_NAMES definition if running on MSYS2.

If this is something you can tackle on this branch before merging, it would be greatly appreciated.

**Comment by Legorooj** [comment]
> Log for `test_spec_with_multipackage[onedir_depends_on_onedir]`:
> [test_multipackage.onedir.log](https://github.com/pyinstaller/pyinstaller/files/4927631/test_multipackage.onedir.log) Any ideas? The `multipackage4_B` folder exists and has everything it needs.

@alexchandel you'll have to try building that test manually - half of the log is missing. Build the spec file yourself and post the output of running the executable.

**Comment by JonathonReinhart** [comment]
@alexchandel Thanks for tackling this head-on!

**Comment by htgoebel** [comment]
Fixing these two failing tests is  major work to be done at the bootloader. This is why there is #4311

**Comment by johnthagen** [comment]
With [Python 3.9 releasing in less than two months](https://www.python.org/dev/peps/pep-0596/), I'm curious if things changed drastically from 3.8 -> 3.9 in terms of the bootloader, etc, or if this PR will more or less get us what's needed for 3.9 support as well?

**Comment by Legorooj** [comment]
@johnthagen we plan for 3.8 and 3.9 support in release 4.1. However, we'll also be dropping python 3.5 (at last - it goes EOL on 13/9/20).

File: PyInstaller/hooks/rthooks/pyi_rth_multiprocessing.py [hunk 8]
```diff
@@ -86,3 +92,27 @@ def __init__(self, *args, **kw):
...
+                    # On some platforms (e.g. AIX) 'os.unsetenv()' is not
+                    # available. In those cases we cannot delete the variable
+                    # but only set it to the empty string. The bootloader
+                    # can handle this case.
+                    if hasattr(os, 'unsetenv'):
+                        os.unsetenv('_MEIPASS2')
+                    else:
+                        os.putenv('_MEIPASS2', '')
+
+    spawning.Popen = _Spawning_Popen
```
[the remark this answers is not in the document]
Comment (Legorooj) [reply PyInstaller/hooks/rthooks/pyi_rth_multiprocessing.py:118]: @reritom are you able to reproduce this?

**Comment by htgoebel** [comment]
Good news: I got some funding for developing this. Expect a release supporting Python 3.8 really soon now.

**Comment by htgoebel** [comment]
For the records: Re. 4a66ebc6ee18c9ba1: I added a hook which prevents heapq to pull in doctest, so this one is not necessary.

**Comment by htgoebel** [comment]
Superseded by 96ad2961d578ce322f3ea80fc9e1fe5d657a1227

Thanks to all for working on it.

**Pull request closed without merging**